### PR TITLE
12.0 pos_product_qty_available

### PR DESCRIPTION
**Pull request by Ommo73**

### CODE REVIEW

**Review by chienandalu — COMMENTED** [review]
Thanks for the contribution :slightly_smiling_face: Is this WIP?

**Comment by Ommo73** [comment]
> Thanks for the contribution  Is this WIP?

its done, but i see some checks failed. should i do something or it can be merged?

**Review by chienandalu — CHANGES_REQUESTED** [review]
Well, first you should start fixing the CI errors. The module can't even be installed :slightly_frowning_face:

**Comment by ilmir-k** [comment]
@Ommo73 
There is the error on runbot:
```
Failed to load registry
Traceback (most recent call last):
  File "/home/odoo/OCB-12.0/odoo/modules/registry.py", line 86, in new
    odoo.modules.load_modules(registry._db, force_demo, status, update_module)
  File "/home/odoo/OCB-12.0/odoo/modules/loading.py", line 421, in load_modules
    loaded_modules, update_module, models_to_check)
  File "/home/odoo/OCB-12.0/odoo/modules/loading.py", line 313, in load_marked_modules
    perform_checks=perform_checks, models_to_check=models_to_check
  File "/home/odoo/OCB-12.0/odoo/modules/loading.py", line 222, in load_module_graph
    load_data(cr, idref, mode, kind='data', package=package, report=report)
  File "/home/odoo/OCB-12.0/odoo/modules/loading.py", line 68, in load_data
    tools.convert_file(cr, package.name, filename, idref, mode, noupdate, kind, report)
  File "/home/odoo/OCB-12.0/odoo/tools/convert.py", line 795, in convert_file
    with file_open(pathname, 'rb') as fp:
  File "/home/odoo/OCB-12.0/odoo/tools/misc.py", line 199, in file_open
    return _fileopen(name, mode=mode, basedir=rtp, pathinfo=pathinfo, basename=basename)
  File "/home/odoo/OCB-12.0/odoo/tools/misc.py", line 257, in _fileopen
    raise IOError('File not found: %s' % basename)
OSError: File not found: pos_product_qty_available/pos_product_qty_available_data.xml
Failed to initialize database `openerp_test`.
Traceback (most recent call last):
  File "/home/odoo/OCB-12.0/odoo/service/server.py", line 1055, in preload_registries
    registry = Registry.new(dbname, update_module=update_module)
  File "/home/odoo/OCB-12.0/odoo/modules/registry.py", line 86, in new
    odoo.modules.load_modules(registry._db, force_demo, status, update_module)
  File "/home/odoo/OCB-12.0/odoo/modules/loading.py", line 421, in load_modules
    loaded_modules, update_module, models_to_check)
  File "/home/odoo/OCB-12.0/odoo/modules/loading.py", line 313, in load_marked_modules
    perform_checks=perform_checks, models_to_check=models_to_check
  File "/home/odoo/OCB-12.0/odoo/modules/loading.py", line 222, in load_module_graph
    load_data(cr, idref, mode, kind='data', package=package, report=report)
  File "/home/odoo/OCB-12.0/odoo/modules/loading.py", line 68, in load_data
    tools.convert_file(cr, package.name, filename, idref, mode, noupdate, kind, report)
  File "/home/odoo/OCB-12.0/odoo/tools/convert.py", line 795, in convert_file
    with file_open(pathname, 'rb') as fp:
  File "/home/odoo/OCB-12.0/odoo/tools/misc.py", line 199, in file_open
    return _fileopen(name, mode=mode, basedir=rtp, pathinfo=pathinfo, basename=basename)
  File "/home/odoo/OCB-12.0/odoo/tools/misc.py", line 257, in _fileopen
    raise IOError('File not found: %s' % basename)
OSError: File not found: pos_product_qty_available/pos_product_qty_available_data.xml
```

File: pos_product_qty_available/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    # "price": 9.00,
+    # "currency": "EUR",
+
+    "depends": [
+        'point_of_sale',
+        'stock',
+    ],
+    "external_dependencies": {"python": [], "bin": []},
+    'data': [
+        'pos_product_qty_available_data.xml',
```
Comment (ilmir-k) [inline pos_product_qty_available/__manifest__.py]: @Ommo73 Use ``data/pos_product_qty_available_data.xml`` instead

**Review by ilmir-k — CHANGES_REQUESTED** [review]
@Ommo73 please fix the travis ci warnings

**Review by Ramil-Mukhametzyanov — COMMENTED** [review]
The POS interface was broken in the new runbot.
And the error appears after the reinstallation the module:

The style compilation failed, see the error below. Your recent actions may be the cause, please try reverting the changes you made.
Could not get content for /pos_product_available/static/src/css/pos.css defined in bundle 'point_of_sale.assets'.

File: pos_product_qty_available/data/pos_product_qty_available_data.xml [hunk 2]
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--Copyright 2014-2016 Ivan Yelizariev <https://it-projects.info/team/yelizariev>
+    Copyright 2018 Gabbasov Dinar <https://it-projects.info/team/GabbasovDinar>
+    Copyright 2018 Kolushov Alexandr <https://it-projects.info/team/KolushovAlexandr>
+    Copyright 2018 Ildar Nasyrov <https://it-projects.info/team/iledarn>
+    License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl.html). -->
+<odoo>
+        <template id="assets_backend" name="pos assets" inherit_id="point_of_sale.assets">
+            <xpath expr="." position="inside">
+                <script type="text/javascript" src="/pos_product_available/static/src/js/pos.js"></script>
```
Comment (ilmir-k) [inline pos_product_qty_available/data/pos_product_qty_available_data.xml]: ``/pos_product_available`` ?
I think it should be ``pos_product_qty_available``

File: pos_product_qty_available/data/pos_product_qty_available_data.xml [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+            </xpath>
+          <xpath expr="//link[@id='pos-stylesheet']" position="after">
+              <link rel="stylesheet" href="/pos_product_available/static/src/css/pos.css"/>
```
Comment (ilmir-k) [inline pos_product_qty_available/data/pos_product_qty_available_data.xml]: the same

File: pos_product_qty_available/data/pos_product_qty_available_data.xml [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+          </xpath>
+        </template>
+        <template id="assets_backend_web" name="pos_product_available_backend_assets" inherit_id="web.assets_backend">
+            <xpath expr="." position="inside">
+                <script type="text/javascript" src="/pos_product_available/static/src/js/test_pos_quantities.js"></script>
```
Comment (ilmir-k) [inline pos_product_qty_available/data/pos_product_qty_available_data.xml]: the same

File: pos_product_qty_available/__manifest__.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
+# Copyright 2014-2018 Ivan Yelizariev <https://it-projects.info/team/yelizariev>
+# Copyright 2017 Gabbasov Dinar <https://it-projects.info/team/GabbasovDinar>
+# Copyright 2018 Kolushov Alexandr <https://it-projects.info/team/KolushovAlexandr>
+# Copyright 2018 Ildar Nasyrov <https://it-projects.info/team/iledarn>
+# License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl.html).
+{
+    "name": """Available quantity of products in POS""",
+    "summary": """Adds available quantity at products in POS""",
+    "category": "Point Of Sale",
+    # "live_test_url": "http://apps.it-projects.info/shop/product/DEMO-URL?version={ODOO_BRANCH}",
```
Comment (chienandalu) [inline pos_product_qty_available/__manifest__.py]: Get rid of commented lines

File: pos_product_qty_available/__manifest__.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    "images": [],
+    "version": "12.0.1.0.0",
+    "application": False,
+
+    "author": "IT-Projects LLC, Ivan Yelizariev, " "Odoo Community Association (OCA)",
```
Comment (chienandalu) [inline pos_product_qty_available/__manifest__.py]: If it fits in the same line it's pointless to have two strings

File: pos_product_qty_available/__manifest__.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    "support": "pos@it-projects.info",
+    "website": "https://github.com/OCA/pos",
+    "license": "LGPL-3",
+    # "price": 9.00,
+    # "currency": "EUR",
```
Comment (chienandalu) [inline pos_product_qty_available/__manifest__.py]: These keys are useless here

File: pos_product_qty_available/__manifest__.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+    "depends": [
+        'point_of_sale',
+        'stock',
+    ],
+    "external_dependencies": {"python": [], "bin": []},
```
Comment (chienandalu) [inline pos_product_qty_available/__manifest__.py]: If no external dependencies are needed, declaring it is useless

File: pos_product_qty_available/__manifest__.py [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    "author": "IT-Projects LLC, Ivan Yelizariev, " "Odoo Community Association (OCA)",
+    "support": "pos@it-projects.info",
+    "website": "https://github.com/OCA/pos",
+    "license": "LGPL-3",
+    # "price": 9.00,
+    # "currency": "EUR",
+
+    "depends": [
+        'point_of_sale',
+        'stock',
```
Comment (chienandalu) [inline pos_product_qty_available/__manifest__.py]: point_of_sale already pulls from stock

File: pos_product_qty_available/__manifest__.py [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+        'data/pos_product_qty_available_data.xml',
+    ],
+    'qweb': [
+        'static/src/xml/pos.xml',
+    ],
+
+    "post_load": None,
+    "pre_init_hook": None,
+    "post_init_hook": None,
+    "uninstall_hook": None,
```
Comment (chienandalu) [inline pos_product_qty_available/__manifest__.py]: Same here, don't declare keys if they aren't needed

File: pos_product_qty_available/__manifest__.py [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+        'stock',
+    ],
+    "external_dependencies": {"python": [], "bin": []},
+    'data': [
+        'data/pos_product_qty_available_data.xml',
+    ],
+    'qweb': [
+        'static/src/xml/pos.xml',
+    ],
+
```
Comment (chienandalu) [inline pos_product_qty_available/__manifest__.py]: And better don't leave blank lines

File: pos_product_qty_available/__init__.py [hunk 12]
```diff
@@ -0,0 +1 @@
+# License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl.html).
```
Comment (chienandalu) [inline pos_product_qty_available/__init__.py]: You can just leave it empty

**Review by chienandalu — CHANGES_REQUESTED** [review]
Please, check lint errors. Some other comments:

**Comment by chienandalu** [comment]
LINT is still wrong: https://travis-ci.org/OCA/pos/jobs/509392075#L607

**Comment by Ommo73** [comment]
> LINT is still wrong: https://travis-ci.org/OCA/pos/jobs/509392075#L607

Can we leave some lines longer than the test requires? This applies to the triggers in the test and lines with the authors.

**Comment by chienandalu** [comment]
The line length warning in JS is non blocking, but you've got a bunch of another errors and warnings: https://travis-ci.org/OCA/pos/jobs/509392075#L720

Most of them can be easily fixed:

- Generate de README.rst file
- Check the blank space warnings
- Check some missing commas, indents, spaces in the eslint

**Comment by chienandalu** [comment]
You're almost there, just fix these flake8 errors: https://travis-ci.org/OCA/pos/jobs/512936427#L603

**Comment by Ommo73** [comment]
@chienandalu flake8 errors fixed

**Review by chienandalu — COMMENTED** [review]
Unblocking this. Reviews needed though

**Comment by osanchezm** [comment]
I am not sure if this is the correct chanel for this question, but I really don't know where to ask. I would like to have a way to show only the products that have quantities available in the POS, or hide the products that have quantities of 0 or less. Do you know if there is some module for that? Or any Idea how to implement? Thanks a lot.

**Comment by ilmir-k** [comment]
> I am not sure if this is the correct chanel for this question, but I really don't know where to ask. I would like to have a way to show only the products that have quantities available in the POS, or hide the products that have quantities of 0 or less. Do you know if there is some module for that? Or any Idea how to implement? Thanks a lot.

@osanchezm contact our managers by it@it-projects.info

**Review by flotho — APPROVED** [review]
Hi,

Tested on runbot, quantity are correctly decremented and displayed.
LGTM

**Review by adnanoscar — APPROVED** [review]

**Review by chienandalu — APPROVED** [review]

**Comment by chienandalu** [comment]
Thanks!

/ocabot merge nobump

**Comment by chienandalu** [comment]
@Ommo73 Looks like test aren't working in the merge. Can you rebase and see?

**Pull request closed without merging**